build: Use a zip instead of dmg for macOS releases

fanquake opened the pull request:
Implements #18128 (mostly). From discussion on that issue we've got 3 `ACK`s (to move from .dmg to .zip), 1 `~0` and 1 `NACK`. So I've opened this PR to give a better overview of what the changes might actually look like. 

Gitian builds currently work, however I haven't tested the signer, and that likely needs some additional changes. As mentioned we can drop quite a few packages from depends. There's some commits here that can cleaned up, squashed or dropped.

If you're compiling macOS, `make appbundle` will (as before) build a dynamically linked `Bitcoin Core.app`. `make deploy` will still build a statically linked `dist/Bitcoin Core.app` and will now also create a `dist/Bitcoin-Core.zip`.

Compiling on Linux there has never been a `make appbundle` rule, and with these changes, `make deploy` now behaves the same as on macOS. The `make deploydir` rule has  been removed. 

Initially I  was unconvinced (of a swap from .dmg to .zip), however I have come around after looking at these changes. If there is not agreement that "this is the way", then I'll revive #18151 and fix it up to include `xorriso` in depends etc.

fanquake committed:
build: remove native_libdmg-hfsplus from depends

fanquake committed:
build: remove native_cdrit from depends

fanquake committed:
build: remove macOS .dmg background image and generation

A make deploy won't currently work because -fancy is still passed to macdeployqtplus

fanquake committed:
build: remove macOS .DS_Store and .plist generation

fanquake committed:
build: add -zip option to macdeployqtplus

This simply zips the staticly built app bundle that has been placed
into /dist.

fanquake committed:
build: note why we set CFBundle*

fanquake committed:
build: update gitian descriptors to use zip method

fanquake committed:
build: remove -dmg and -fancy argument from macdeployqtplus

fanquake committed:
build: remove -sign argument from macdeployqtplus

fanquake committed:
build: merge -zip and -volname arguments in macdeployqtplus

fanquake committed:
build: move qt_conf to where it's actually used in macdeployqtplus

fanquake committed:
build: remove -add-resources argument from macdeployqtplus

fanquake committed:
docs: update macOS docs for new zip deployment method

ollitsac107 commented:
El mar., 26 de mayo de 2020 2:55 a. m., Michael <notifications@github.com>
escribió:

> Implements #18128 <https://github.com/bitcoin/bitcoin/issues/18128>
> (mostly). From discussion on that issue we've got 3 ACKs (to move from
> .dmg to .zip), 1 ~0 and 1 NACK. So I've opened this PR to give a better
> overview of what the changes might actually look like.
>
> Gitian builds currently work, however I haven't tested the signer, and
> that likely needs some additional changes. As mentioned we can drop quite a
> few packages from depends. There's some commits here that can cleaned up,
> squashed or dropped.
>
> If you're compiling macOS, make appbundle will (as before) build a
> dynamically linked Bitcoin Core.app. make deploy will still build a
> statically linked dist/Bitcoin Core.app and will now also create a
> dist/Bitcoin-Core.zip.
>
> Compiling on Linux there has never been a make appbundle rule, and with
> these changes, make deploy now behaves the same as on macOS. The make
> deploydir rule has been removed.
>
> Initially I was unconvinced (of a swap from .dmg to .zip), however I have
> come around after looking at these changes. If there is not agreement that
> "this is the way", then I'll revive #18151
> <https://github.com/bitcoin/bitcoin/pull/18151> and fix it up to include
> xorriso in depends etc.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/bitcoin/bitcoin/pull/19068
> Commit Summary
>
>    - build: add -zip option to macdeployqtplus
>    - build: remove native_libdmg-hfsplus from depends
>    - build: remove native_cdrit from depends
>    - build: remove macOS .dmg background image and generation
>    - build: remove macOS .DS_Store and .plist generation
>    - build: note why we set CFBundle*
>    - build: update gitian descriptors to use zip method
>    - build: remove -dmg and -fancy argument from macdeployqtplus
>    - build: remove -sign argument from macdeployqtplus
>    - build: merge -zip and -volname arguments in macdeployqtplus
>    - build: move qt_conf to where it's actually used in macdeployqtplus
>    - build: remove -add-resources argument from macdeployqtplus
>    - docs: update macOS docs for new zip deployment method
>
> File Changes
>
>    - *M* Makefile.am
>    <https://github.com/bitcoin/bitcoin/pull/19068/files#diff-c949f93d03f44a4217d7a138f9e2e54a>
>    (65)
>    - *M* ci/test/00_setup_env_mac.sh
>    <https://github.com/bitcoin/bitcoin/pull/19068/files#diff-f8f760195adbee096a0a78c72242bce4>
>    (2)
>    - *M* configure.ac
>    <https://github.com/bitcoin/bitcoin/pull/19068/files#diff-67e997bcfdac55191033d57a16d1408a>
>    (6)
>    - *M* contrib/gitian-descriptors/gitian-osx-signer.yml
>    <https://github.com/bitcoin/bitcoin/pull/19068/files#diff-e93eadfeb0c10fed75cd1434371bb41c>
>    (8)
>    - *M* contrib/gitian-descriptors/gitian-osx.yml
>    <https://github.com/bitcoin/bitcoin/pull/19068/files#diff-2bf50096db55445178e7238c402f2e5c>
>    (19)
>    - *M* contrib/macdeploy/README.md
>    <https://github.com/bitcoin/bitcoin/pull/19068/files#diff-9de36befe13356841c2699ee0eff4a0a>
>    (45)
>    - *D* contrib/macdeploy/background.svg
>    <https://github.com/bitcoin/bitcoin/pull/19068/files#diff-8552c751b30ae0c0bbb7a7994e13a986>
>    (34)
>    - *D* contrib/macdeploy/custom_dsstore.py
>    <https://github.com/bitcoin/bitcoin/pull/19068/files#diff-b7d8f8738f97ba32c64050567a536ab4>
>    (59)
>    - *D* contrib/macdeploy/fancy.plist
>    <https://github.com/bitcoin/bitcoin/pull/19068/files#diff-c5ee10fdc4e723d0e3bc04985a9b0a88>
>    (32)
>    - *M* contrib/macdeploy/macdeployqtplus
>    <https://github.com/bitcoin/bitcoin/pull/19068/files#diff-e2c1caf49768933aa4719737154e63c8>
>    (275)
>    - *M* depends/README.md
>    <https://github.com/bitcoin/bitcoin/pull/19068/files#diff-36bb097207a4bca436880cdc4cd3f66b>
>    (2)
>    - *D* depends/packages/native_biplist.mk
>    <https://github.com/bitcoin/bitcoin/pull/19068/files#diff-1dd190bdf3a4e6a2d8035444e4517948>
>    (15)
>    - *D* depends/packages/native_cdrkit.mk
>    <https://github.com/bitcoin/bitcoin/pull/19068/files#diff-84e75f47808eba6dd2903afee0156b54>
>    (26)
>    - *D* depends/packages/native_ds_store.mk
>    <https://github.com/bitcoin/bitcoin/pull/19068/files#diff-6d3cefe7abc09f03958627a7c0e9256d>
>    (16)
>    - *D* depends/packages/native_libdmg-hfsplus.mk
>    <https://github.com/bitcoin/bitcoin/pull/19068/files#diff-70c3d4e88ebfdac4940b759ec8d4e35f>
>    (24)
>    - *D* depends/packages/native_mac_alias.mk
>    <https://github.com/bitcoin/bitcoin/pull/19068/files#diff-1fdc24ec38b3c7b6d35d21494e93a296>
>    (15)
>    - *M* depends/packages/packages.mk
>    <https://github.com/bitcoin/bitcoin/pull/19068/files#diff-d7aa5d84559a1c436fb56a7def7f98c0>
>    (4)
>    - *D* depends/patches/native_cdrkit/cdrkit-deterministic.patch
>    <https://github.com/bitcoin/bitcoin/pull/19068/files#diff-fc6091667acaa99f573220e16a7bc06b>
>    (86)
>    - *D*
>    depends/patches/native_libdmg-hfsplus/remove-libcrypto-dependency.patch
>    <https://github.com/bitcoin/bitcoin/pull/19068/files#diff-92b035867966adf9ac5dff978c7c4096>
>    (45)
>    - *M* doc/build-osx.md
>    <https://github.com/bitcoin/bitcoin/pull/19068/files#diff-5c9db1d4619895b002078a3529a8dc72>
>    (13)
>    - *M* doc/dependencies.md
>    <https://github.com/bitcoin/bitcoin/pull/19068/files#diff-0c8311709d11060c5156268e58f5f209>
>    (4)
>
> Patch Links:
>
>    - https://github.com/bitcoin/bitcoin/pull/19068.patch
>    - https://github.com/bitcoin/bitcoin/pull/19068.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/19068>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AO23BJAVI7UXVROZA2C5PHTRTNYYNANCNFSM4NKCADEQ>
> .
>

jonasschnelli commented:
Gitian Build seems to work: https://bitcoin.jonasschnelli.ch/gitian/build/142 
Maybe the ZIP filename could still contain the phrase "macOS" or "osx".

Conceptual, I'm currently ~0 (I [commented](https://github.com/bitcoin/bitcoin/issues/18128#issuecomment-585323098) on the concept before).
I think it is nice to simplify the build system (probably makes it much easier for guix). Though, I'm currently unsure if this slightly worsens the UX for macOS users. If it does, should we reduce the UX to simplify the build-system (is that an acceptable tradeoff)?

However, if there is a difference in user experience between a zip and a dmg, it is probably marginal.

MarcoFalke commented:
Is there any "official" mac documention or user surveys that suggest which one is the recommended or preferred way?

dongcarl commented:
Concept ACK! Great to see the amount of code we can remove.

-----

> (probably makes it much easier for guix).

Just to make myself clear, I am _**for**_ this change because it is a great simplification for a marginal difference in user experience and not because of anything specific to Guix (my current implementation of the Guix macOS build is already able to build the `dmg`). :relaxed:

Sjors commented:
You can add a symlink to `/Applications` to the archive and call it "Drag Here".
<img width="351" alt="Schermafbeelding 2020-05-27 om 18 29 17" src="https://user-images.githubusercontent.com/10217/83046935-02020e00-a048-11ea-9d3a-1f0975cfed9c.png">

With that it's close enough to what macOS users are used to, with a nice reduction in dependencies and complexity.

You could also keep the `.DS_Store` and `.background` folder around for full branding and a nicer layout, but that's pretty marginal.

fanquake commented:
> You could also keep the .DS_Store and .background folder around for full branding and a nicer layout, but that's pretty marginal.

"Keeping that around" would mostly defeat the point of these changes.

Sjors commented:
> "Keeping that around" would mostly defeat the point of these changes.

It still gets rid of `genisoimage` and `libdmg-hfsplus`, but indeed `.DS_Store` and `.background` require a lof of tools themselves. Just the symlink should get us close enough to the original UX.

fanquake commented:
Pushed a few more removals, updated the descriptors and fixed whitespace.

> Is there any "official" mac documention or user surveys that suggest which one is the recommended or preferred way?

The Apple documentation is mainly focused on using Xcode and App Store distribution: https://help.apple.com/xcode/mac/current/#/devac02c5ab8. There is a section on distribution outside the App Store: https://help.apple.com/xcode/mac/current/#/dev033e997ca, however I couldn't quickly find any mention of using `.dmg`s. There are some instructions on "macOS installer packages" (`.pkg`): https://help.apple.com/xcode/mac/current/#/deve51ce7c3d.

There seem to be at least some companies that are using a combination of `.dmg` and `.zip` for distribution. i.e: [Sublime Merge](https://www.sublimemerge.com/) is distributed as a `.zip`, while [Sublime Text](https://www.sublimetext.com/) is a `.dmg`.

> Just the symlink should get us close enough to the original UX.

I'm not aware of any projects that are using a symlink inside a `.zip` approach. The Bitcoin Core GUI is mostly [designed for power users of bitcoin](https://github.com/bitcoin/bitcoin/issues/17395#issuecomment-550339984), and I think the intersection of power users, and people who don't know how to handle a `.zip` file would be very small. 

Also, as far as I'm aware, you can open and use an app bundle from any directory on macOS, and it makes  little to no difference to how the app actually runs?

Sjors commented:
> I'm not aware of any projects that are using a symlink inside a .zip approach

That's because most projects use a DMG with a symlink and background image. We have a good reason to avoid that, namely to make deterministic builds easier. Even as an experienced macOS user, I find the symlink useful; it saves me from opening a new Finder window, navigating to Applications and then dragging between two windows. And it's a trivial thing to add.

> Also, as far as I'm aware, you can open and use an app bundle from any directory on macOS, and it makes little to no difference to how the app actually runs?

Whenever I deal with applications that don't have an installer or a symlink, I very often make the mistake of running it from the Download folder. That works, until I accidentally clean up the download folder. There's no obvious alternative to drag the program to.

fanquake commented on Makefile.am:
```diff
@@ -116,50 +111,18 @@ osx_volname:
 	echo $(OSX_VOLNAME) >$@
 
 if BUILD_DARWIN
-$(OSX_DMG): $(OSX_APP_BUILT) $(OSX_PACKAGING) $(OSX_BACKGROUND_IMAGE)
```
One possible simplification after this change, would be to make the `otool` and `installnametool` checks no longer cross-compiling dependent. Then we should be able to have a single `OSX_ZIP` rule, and drop the need for `BUILD_DARWIN`.

luke-jr commented:
Can we add the ZIP and remove the DMG in two steps?

IMO it'd be nice to build both for 0.21, but only distribute the ZIP. That way if it causes too much problems/confusion we can just change it back without rebuilds/etc.

fanquake commented:
I might come back to this in future, but closing for now as I've made some changes to the existing macdeploy process. Any discussion should continue in #18128.
